### PR TITLE
Py3 testing cleanups

### PR DESCRIPTION
**Pull request by nsoranzo**
Small follow-ups on the amazing work of @mvdbeek in #6561, #6565,  #6595 and #6605.

See individual commits for details.

xref. #1715

### CODE REVIEW

File: lib/galaxy/tools/evaluation.py [hunk 1]
```diff
@@ -575,8 +574,8 @@ def __write_workdir_file(self, config_filename, content, context, is_template=Tr
             value = fill_template(content, context=context)
         else:
             value = content
-        with open(config_filename, "wb") as f:
-            f.write(smart_str(value))
+        with open(config_filename, "w") as f:
```
Comment (mvdbeek) [inline lib/galaxy/tools/evaluation.py]: This was my initial attempt, but that broke the unicode test tool on docker. I might have changed the encoding in the image in the meantime though, I'll double check and point jenkins at your branch.

File: test/base/populators.py [hunk 2]
```diff
@@ -24,10 +25,10 @@ def nottest(x):
 
 
 # Simple workflow that takes an input and call cat wrapper on it.
-workflow_str = resource_string(__name__, "data/test_workflow_1.ga").decode('utf-8')
+workflow_str = unicodify(resource_string(__name__, "data/test_workflow_1.ga"))
```
Comment (mvdbeek) [inline test/base/populators.py:28]: I guess you'll want to convert them all, there are more in 
```
lib/galaxy/exceptions/error_codes.py
45:error_codes_json = resource_string(__name__, 'error_codes.json').decode("UTF-8")

lib/galaxy/jobs/runners/util/job_script/__init__.py
16:    resource_string(__name__, 'DEFAULT_JOB_FILE_TEMPLATE.sh').decode('UTF-8')
20:    resource_string(__name__, 'CLUSTER_SLOTS_STATEMENT.sh').decode('UTF-8')
23:    resource_string(__name__, 'MEMORY_STATEMENT.sh').decode('UTF-8')
```

File: lib/galaxy/web/stack/__init__.py [hunk 3]
```diff
@@ -299,8 +299,9 @@ def _configured_farms(self):
         if self._farms_dict is None:
             self._farms_dict = {}
             farms = uwsgi.opt.get('farm', [])
-            farms = farms if isinstance(farms, list) else [unicodify(farms)]
+            farms = farms if isinstance(farms, list) else [farms]
             for farm in farms:
+                farm = unicodify(farms)
```
Comment (mvdbeek) [inline lib/galaxy/web/stack/__init__.py]: This should be `farm = unicodify(farm)`

File: lib/galaxy/tools/evaluation.py (hunk 1 above)
Comment (mvdbeek) [reply lib/galaxy/tools/evaluation.py]: yeah, this still fails in the docker image:
```
FAIL
( unicode_stream ) > Test-3 ... galaxy.tools DEBUG 2018-08-21 07:44:42,132 Validated and populated state for tool request (0.047 ms)
galaxy.tools.actions INFO 2018-08-21 07:44:42,147 Handled output named out_file1 for tool unicode_stream (0.587 ms)
galaxy.tools.actions INFO 2018-08-21 07:44:42,149 Added output datasets to history (1.813 ms)
galaxy.tools.actions INFO 2018-08-21 07:44:42,150 Verified access to datasets for Job[unflushed,tool_id=unicode_stream] (0.026 ms)
galaxy.tools.actions INFO 2018-08-21 07:44:42,151 Setup for job Job[unflushed,tool_id=unicode_stream] complete, ready to flush (1.757 ms)
galaxy.tools.actions INFO 2018-08-21 07:44:42,166 Flushed transaction for job Job[id=3,tool_id=unicode_stream] (15.094 ms)
galaxy.tools.execute DEBUG 2018-08-21 07:44:42,167 Tool [unicode_stream] created job [3] (23.101 ms)
galaxy.tools.execute DEBUG 2018-08-21 07:44:42,171 Executed 1 job(s) for tool unicode_stream request: (38.651 ms)
galaxy.jobs.mapper DEBUG 2018-08-21 07:44:42,516 (3) Mapped job to destination id: local
galaxy.jobs.handler DEBUG 2018-08-21 07:44:42,520 (3) Dispatching to local runner
galaxy.jobs DEBUG 2018-08-21 07:44:42,524 (3) Working directory for job is: /tmp/tmp76q95lht/tmpo0l6a5id/tmpvh6dmpqm/database/job_working_directory_ico6qlt6/000/3
galaxy.jobs DEBUG 2018-08-21 07:44:42,526 (3) Persisting job destination (destination id: local)
galaxy.jobs.runners DEBUG 2018-08-21 07:44:42,534 Job [3] queued (13.951 ms)
galaxy.jobs.handler INFO 2018-08-21 07:44:42,540 (3) Job dispatched
galaxy.jobs.runners ERROR 2018-08-21 07:44:42,582 (3) Failure preparing job
Traceback (most recent call last):
  File "/galaxy/lib/galaxy/jobs/runners/__init__.py", line 187, in prepare_job
    job_wrapper.prepare()
  File "/galaxy/lib/galaxy/jobs/__init__.py", line 874, in prepare
    self.command_line, self.extra_filenames, self.environment_variables = tool_evaluator.build()
  File "/galaxy/lib/galaxy/tools/evaluation.py", line 435, in build
    raise e
  File "/galaxy/lib/galaxy/tools/evaluation.py", line 431, in build
    self.__build_config_files()
  File "/galaxy/lib/galaxy/tools/evaluation.py", line 504, in __build_config_files
    self.__write_workdir_file(config_filename, config_text, param_dict, is_template=is_template)
  File "/galaxy/lib/galaxy/tools/evaluation.py", line 578, in __write_workdir_file
    f.write(value)
UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-5: ordinal not in range(128)
```

**Comment by mvdbeek** [comment]
I've pointed jenkins are your branch, but you still have to trigger manually

File: lib/galaxy/tools/evaluation.py (hunk 1 above)
Comment (mvdbeek) [reply lib/galaxy/tools/evaluation.py]: (https://jenkins.galaxyproject.org/job/docker-framework-py3/38/testReport/)

File: lib/galaxy/web/stack/__init__.py (hunk 3 above)
Comment (nsoranzo) [reply lib/galaxy/web/stack/__init__.py]: Fixed, thanks!

File: test/base/populators.py (hunk 2 above)
Comment (nsoranzo) [reply test/base/populators.py:28]: Fixed those too, thanks!

File: lib/galaxy/tools/evaluation.py (hunk 1 above)
Comment (nsoranzo) [reply lib/galaxy/tools/evaluation.py]: I've added a debug statement and found out that `f.encoding` is `ANSI_X3.4-1968`  when run through Docker.

Comment (nsoranzo) [reply lib/galaxy/tools/evaluation.py]: Fixed using `io.open(config_filename, "w", encoding='utf-8')` .

File: test/functional/twilltestcase.py [hunk 4]
```diff
@@ -1,443 +0,0 @@
-from __future__ import print_function
```
Comment (jmchilton) [inline test/functional/twilltestcase.py:1]: ❤️

**Review by mvdbeek — APPROVED** [review]

**Comment by mvdbeek** [comment]
Awesome, thanks a lot @nsoranzo !

**Comment by nsoranzo** [comment]
For the record, Py3 Jenkins builds were green too:
api tests on Py3: https://jenkins.galaxyproject.org/job/docker-api-py3/125/
framework tests on Py3: https://jenkins.galaxyproject.org/job/docker-framework-py3/42/
integration tests on Py3: https://jenkins.galaxyproject.org/job/docker-integration-py3/50/

**Comment by jmchilton** [comment]
Wonderful, thanks a bunch!

**Comment by mvdbeek** [comment]
Yep, and we now have the triggerphrase `galaxybot test py3` if you want to test a PR

**Comment by nsoranzo** [comment]
> Yep, and we now have the triggerphrase galaxybot test py3 if you want to test a PR

Nice!

**Pull request merged**